zerver/tests/test_subs.py

```diff
@@ -5115,7 +5115,7 @@ def test_subscriptions_query_count(self) -> None:
         notifications_stream = get_stream(self.streams[0], self.test_realm)
         self.test_realm.notifications_stream_id = notifications_stream.id
         self.test_realm.save()
-        with self.assert_database_query_count(45):
+        with self.assert_database_query_count(44):
```

Comment:
Why does this change save a database query? Obviously not a bad thing, but I don't understand the mechanism.

Reply:
Ran the migration with SQL output for this test case, the removed query was:
```sql
SELECT "zerver_realmfilter"."id", 
  "zerver_realmfilter"."realm_id",
  "zerver_realmfilter"."pattern",
  "zerver_realmfilter"."url_format_string"
FROM "zerver_realmfilter" WHERE "zerver_realmfilter"."realm_id" = 2
```
I'm sure this query was invoked somewhere inside `add_subscriptions_backend`. This query is skipped probably because of some ORM caching logic. It can be verified that this call to `render_stream_description` is relevant: https://github.com/zulip/zulip/blob/8023a646def9c9e371871e48d548649225b1e1a2/zerver/lib/streams.py#L158-L159
(where we add `realm` as the second argument in this PR)

Reply:
I see, this makes a lot of sense.

Digging back up in the call stack, I assume that since `message_realm` is no longer `None` now when `do_convert` is called via `render_realm_description`, we get to call `maybe_update_markdown_engines` that populates the linkifier cache with a linkifier key for the realm.
```
def do_convert(
    content: str,
    realm_alert_words_automaton: Optional[ahocorasick.Automaton] = None,
    message: Optional[Message] = None,
    message_realm: Optional[Realm] = None,
    sent_by_bot: bool = False,
    translate_emoticons: bool = False,
    url_embed_data: Optional[Dict[str, Optional[UrlEmbedData]]] = None,
    mention_data: Optional[MentionData] = None,
    email_gateway: bool = False,
    no_previews: bool = False,
) -> MessageRenderingResult:
    """Convert Markdown to HTML, with Zulip-specific settings and hacks."""
    # This logic is a bit convoluted, but the overall goal is to support a range of use cases:
    # * Nothing is passed in other than content -> just run default options (e.g. for docs)
    # * message is passed, but no realm is -> look up realm from message
    # * message_realm is passed -> use that realm for Markdown purposes
    if message is not None:
        if message_realm is None:
            message_realm = message.get_realm()
    if message_realm is None:
        linkifiers_key = DEFAULT_MARKDOWN_KEY
    else:
        linkifiers_key = message_realm.id

    if message and hasattr(message, "id") and message.id:
        logging_message_id = "id# " + str(message.id)
    else:
        logging_message_id = "unknown"

    if message is not None and message_realm is not None:
        if message_realm.is_zephyr_mirror_realm:
            if message.sending_client.name == "zephyr_mirror":
                # Use slightly customized Markdown processor for content
                # delivered via zephyr_mirror
                linkifiers_key = ZEPHYR_MIRROR_MARKDOWN_KEY

    maybe_update_markdown_engines(linkifiers_key, email_gateway)
```